store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/util/MergeSchema.java

```diff
@@ -48,20 +60,52 @@ public Schema apply(final Schema update, final Schema state) {
+            final String visibilityProperty = (boolean) context.getOrDefault(WIPE_VISIBILITY_PROPERTY, false)
+                    ? null
+                    : update.getVisibilityProperty();
+
+            Schema.Builder mergeSchema = new Schema.Builder(state);
+
             try {
-                mergeSchema.merge(new Schema.Builder(update).vertexSerialiser(vertexSerialiser).build());
-            } catch (final SchemaException e) {
-                if (e.getMessage().contains(Schema.UNABLE_TO_MERGE_SCHEMAS_CONFLICT_WITH_VERTEX_SERIALISER_OPTIONS_ARE)) {
-                    LOGGER.error(String.format("Caught SchemaException, attempting to re-merge but without vertex serialisers. Error message:%s", e.getMessage()));
+                mergeSchema.merge(new Schema.Builder(update)
+                        .vertexSerialiser(vertexSerialiser)
+                        .visibilityProperty(visibilityProperty)
+                        .build());
+
+            } catch (final Exception e) {
+                if (e.getMessage().contains(SCHEMAS_CONFLICT_WITH_VERTEX_SERIALISER)) {
+                    LOGGER.error(String.format(FORMAT_CAUGHT_SCHEMA_EXCEPTION_ATTEMPTING_TO_RE_MERGE_BUT_WITHOUT_S_ERROR_MESSAGE_S, VERTEX_SERIALISER, e.getMessage()));
                     //Clashing Vertex Serialiser is possibly a recoverable state, continue without using Vertex Serialisers, retain this state.
                     context.put(WIPE_VERTEX_SERIALISERS, true);
-                    mergeSchema.merge(new Schema.Builder(update).vertexSerialiser(null).build());
+                    mergeSchema.merge(new Schema.Builder(update)
+                            .vertexSerialiser(null)
+                            .build());
                     mergeSchema.vertexSerialiser(null);
+                } else if (e.getMessage().contains(SCHEMAS_CONFLICT_WITH_VISIBILITY_PROPERTY)) {
+                    LOGGER.error(String.format(FORMAT_CAUGHT_SCHEMA_EXCEPTION_ATTEMPTING_TO_RE_MERGE_BUT_WITHOUT_S_ERROR_MESSAGE_S, VISIBILITY_PROPERTY, e.getMessage()));
+                    //Clashing visibility property is possibly a recoverable state, continue without using visibility property, retain this state.
+                    context.put(WIPE_VISIBILITY_PROPERTY, true);
+                    mergeSchema.merge(new Schema.Builder(update)
+                            .visibilityProperty(null)
+                            .build());
+                    mergeSchema.visibilityProperty(null);
+                } else if (e.getMessage().contains(SCHEMAS_CONFLICT_WITH_TYPE_CLASS)
+                        || e.getMessage().contains(SCHEMAS_CONFLICT_WITH_TYPE_SERIALISER)
+                        || e.getMessage().contains(SCHEMAS_CONFLICT_WITH_AGGREGATE_FUNCTION)) {
+                    // Not possible to resolve from these collisions.
+                    throw new SchemaException(String.format(FORMAT_MERGE_FUNCTION_UNABLE_TO_RECOVER_FROM_ERROR_DUE_TO_S, e.getMessage()), e);
```

Comment:
The error message is more informative for the error on line 105.
Yeah there was a hope to be able to fix this, but it's currently not possible without new feature and behaviour requests. Altho verbose I thought it was good for this code to remain, to demonstrate that is specific state that can't be recovered.

maybe I change error message on line 108 to something like that does NOT say "unable to recover" because it is a catch all, the exception isn't as expected as the others?

🤷‍♂️

Reply:
Yes, a different error message would be good. This would not be duplicating anything then.

Reply:
done